Additions/XCTestCase-KIFAdditions.m

```diff
@@ -65,6 +65,17 @@ - (void)KIF_recordUnexpectedFailureWithDescription:(id)arg1 exception:(NSExcepti
 - (void)writeScreenshotForException:(NSException *)exception;
 {
     [[UIApplication sharedApplication] writeScreenshotForLine:[exception.userInfo[@"LineNumberKey"] unsignedIntegerValue] inFile:exception.userInfo[@"FilenameKey"] description:nil error:NULL];
+    
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    [XCTContext runActivityNamed:(@"screenshot") block:^(id<XCTActivity>  _Nonnull activity) {
+        XCUIScreenshot *screenShot = [[XCUIScreen mainScreen] screenshot];
```

Comment:
This requires having the `UIAutomation` framework linked into KIF. We do have this today, but I'm hesitant to add more dependencies on `XCUI`, which we can't use on devices.

At the very least, this needs to be wrapped in a `#if TARGET_IPHONE_SIMULATOR` like the method `acknowledgeSystemAlert` is (for the same reasons).

Reply:
thats right, I will be adding those #if s

Reply:
apparantly  the above code for write screenshot doest work on device too and it will crash on 
```             *error = [NSError KIFErrorWithFormat:@"Couldn't create directory at path %@ (details: %@)", outputPath, directoryCreationError];
```  so it doesn't reach my code. 
so if that doesnt work on device I should be checking for target for that one too , am I right ?

Reply:
actually I didn't quit get what you mean by `having the UIAutomation framework linked into KIF`.  I tried on device and it worked fine in my project.

Reply:
Take a look at `-[UIAutomationHelper linkAutomationFramework]` for what I mean. The screenshotting method (and all things relying on XCUI) should probably be exposed as methods from that class.

Reply:
but I do want to have it for devices as well. cause it can be used on devices.  what you are saying is that instead of checking for simulator I can put the screenshot codes in the `UIAutomationHelper` class as a function and call that to get the screenshot  in `XCTestCae-KIFAdditions` ?

Reply:
Yep, that's what I meant for where to put this code.

Regarding having this functionality work on physical iOS devices, unfortunately it can't (except jailbroken devices) with this relying on XCUI for the functionality. This is the portion that will fail:
https://github.com/kif-framework/KIF/blob/2453685d15221a50ba73bc345031d940e864e7be/Classes/UIAutomationHelper.m#L193

It might be worthwhile to file a radar to Apple requesting them to expose this functionality for usage within XCTest.

Reply:
This should be moved to `UIAutomationHelper`, and all public functionality of it should probably be made to either:

1. be compiled out on physical devices (fail at compile time, don't support jailbroken devices)
2. crash on calling one of these methods on a physical device (fail at runtime, support jailbroken devices)
3. store a bool if the call to `linkAutomationFramework` fails and log some message (degrade gracefully, less obvious failure mode)

Reply:
Thanks for explanation. I agree that I should move those UI related things to that class and I know that the `acknowledgeSystemAlert` is not supported for actual device except the jailbroken ones. but here in my case it does actually work on non-jailbroken device. even the UIAutomation is disabled it will get the screenshot of the device. so I don't get your point about handling jailbroken and non-jailbroken device ??

Reply:
You're able to run this on a real physical iOS device, running a KIF test as a unit test (not as UI test)? That is surprising to me. I'll have to give this a shot, but I don't generally run our automation on physical devices. I'll try to verify this soon.

Reply:
Yes exactly ! KIF as unit test. as I understand these code are all in XCTest framework so we don't have those issue that you are talking about.  it will run fine on simulator and real physical device ( not jailbroken ) !